### PR TITLE
Adding Type Hints

### PR DESCRIPTION
**Pull request by craftamap**
## Description of Change

When developing Python3 in a IDE, tooling support like auto completion/intellisense is only limited due to pythons dynamic typing. This often leads to situations where the source code of a function somebody wants to use needs the be read first in order to understand the return value. While this can be optimised by documenting the return type in a python docstring, the proper solution since python 3.6 is to so called [type hints](https://www.python.org/dev/peps/pep-0484/). While not being evaluated during runtime, type hints are very useful during development, as many IDEs can interprete those and give proper auto completion. This improves the overall development experience. With the built-in library `typing`, even the type of list-items and dictionarys (and many more) can be hinted.

See a demo of the behaviour in visual studio code:

### Without type hints

Current header of the `validate_and_resolve_descriptor`-function:

Screenshot_20201125_005755

Usage of the return value:

image
 
Nothing is recommended to the user

### With type hints

New header of the `validate_and_resolve_descriptor`-function:

image

Usage of the return value:

image

As you can see, type specific recommendations are recognized by the IDE.

This pull request proposed the introduction of described type hints into this project by creating type hints for the first two files. If the project wants to use type hints, feel free to contact me, as I will create them for all currently existing classes. :)

## Fixes

No issue was created for this

## Did you test your changes?
- [X] Yes

## Reviewers
@seanmarpo

### CODE REVIEW

**Comment by seanmarpo** [comment]
Thanks for bringing this up - I completely agree that using typing is the right way to go This whole tool was mostly written in a single evening for an internal Hackathon, so adding type hints wasn't my first priority.

If you'd like to go through the hassle of adding typing information, by all means, feel free. Otherwise, I will throw this on my personal backlog.

I personally like https://github.com/facebook/pyre-check to do type checking, so I would probably suggest adding in a Github actions check to run pyre once completed.

Again, no pressure - I'm happy to clean up the code as well, but I truly appreciate the help.

**Comment by craftamap** [comment]
I didn't even now of pyre-check - looks like cool addition :)

For now, I will go through all files and add the existing types. I feel like this is a good work I can do while waiting for other stuff to finish. :)

**Comment by craftamap** [comment]
@seanmarpo I added type hints to all files - consider this pull request ready for review!

**Comment by seanmarpo** [comment]
Thanks for doing this. Since the tests make some rather sketchy network calls, I have reason to believe Github is blocking them. I'll review this and fight with the tests at the same time. 👍

File: scans/descriptor_scan.py [hunk 1]
```diff
@@ -51,23 +52,23 @@ def _get_links(self):
...
 
-    def _fill_context_vars(self, url):
+    def _fill_context_vars(self, url: str) -> str:
         url = re.sub(CONDITION_MATCHER, 'true', url, flags=re.IGNORECASE)
         url = re.sub(BRACES_MATCHER, 'test', url)
 
         return url
 
-    def _find_urls_in_module(self, module):
+    def _find_urls_in_module(self, module: dict) -> typing.List[str]:
```
Comment (seanmarpo) [inline scans/descriptor_scan.py:68]: I believe this should be `module: Union[dict, list]`

File: scans/tls_scan.py [hunk 2]
```diff
@@ -95,7 +97,7 @@ def _get_hsts_info(self, scan_res):
 
         return hsts_present
 
-    def scan(self):
+    def scan(self) -> TlsResult:
         logging.info(f"Starting SSL/TLS Scan for {self.domain}...")
         server_info = self._check_connectivity()
         scan_res = self._run_scan(server_info)
```
Comment (seanmarpo) [inline scans/tls_scan.py:103]: Type checker reports an error here -- Kinda funny. It found an edge case that I hadn't considered. If `server_info` is None, an exception will be thrown. I'll tag this as a bug for me to fix.

**Review by seanmarpo — COMMENTED** [review]
General comment - I'd prefer to we import the specific classes from Typing so the type annotations aren't as verbose.

eg., `from typing import Union, List, Optional....` so we can then just tag things as such without needing the leading `typing.` on everything.

**Comment by craftamap** [comment]
>  I'd prefer to we import the specific classes from Typing so the type annotations aren't as verbose. 

Sure, will update later!

**Comment by seanmarpo** [comment]
Hey @craftamap - I went ahead and made the changes on your branch, but I don't have perms to to push it up to this PR. Can you double-check that you've allowed edits from maintainers? Thanks!

**Comment by craftamap** [comment]
@seanmarpo Huh, thats wierd...

image

Anyway, I sent you an invite to contribute to my fork. If you accept it, it should be possible in any case :)

**Comment by seanmarpo** [comment]
Weird, it worked now. Probably my fault.

Looks like Github actions is having a partial outage, so I'll ping this thread when CI/CD runs. The commit was added. Please review and let me know if you're okay with my changes. Thanks.

**Review by craftamap — COMMENTED** [review]
As I can't approve my own pr: your changes look good. Just one minor thing :)

File: models/requirements.py [hunk 3]
```diff
@@ -8,7 +8,7 @@ class RequirementsResult(JsonObject):
     title = StringProperty()
     proof = ListProperty(StringProperty())
 
-    def was_scanned(self):
+    def was_scanned(self) -> bool:
```
[the remark this answers is not in the document]
Comment (seanmarpo) [reply models/requirements.py:11]: Wow - Fantastic catch. I had no idea Python handled booleans this way. Fixed. It should only return a bool.

**Review by seanmarpo — APPROVED** [review]

**Comment by craftamap** [comment]
Nice! Is there anything left to do? :)

**Pull request merged**